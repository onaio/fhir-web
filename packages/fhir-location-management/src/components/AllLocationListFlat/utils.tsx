import type { IBundle } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/IBundle';
import { BundleEntry } from '@smile-cdr/fhirts/dist/FHIR-R4/classes/bundleEntry';
import { Resource } from '@smile-cdr/fhirts/dist/FHIR-R4/classes/resource';
import { Dictionary } from '@onaio/utils';
import { MATCH_SEARCH_MODE } from '../../constants';
import { PhysicalTypeCodes, hl7PhysicalTypeCodeSystemUri } from '@opensrp/fhir-helpers';

/**
 * @param bundle - a fhir resource bundle api response
 */
export function getEntryFromBundle<TEntry>(bundle: IBundle) {
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  const entry = bundle.entry?.filter((x) => x !== undefined) || [];
  return entry as TEntry[];
}

/**
 * @param resource - a fhir resource
 * @param resourcesById - a Dictionary fhir resource by id
 */
export function getResourceParentName(
  resource: Dictionary,
  resourcesById: Dictionary<Resource>
): string {
  const partOfRef = resource.partOf?.reference;
  const partOfDisplay = resource.partOf?.display;
  if (!partOfRef) {
    return '';
  }
  const parentId = partOfRef.split('/')[1];
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  const parentResource: Dictionary = resourcesById[parentId] ?? {};
  const parentName = parentResource.name;
  return parentName || partOfDisplay || '';
}

/**
 * @param data - a fhir resource bundle entry
 */
export function getTableData(data: BundleEntry[]) {
  const resourcesById: Dictionary<Resource> = {};
  const tableData: Dictionary<string | Dictionary>[] = [];
  data.forEach((entry) => {
    const id = entry.resource?.id;
    if (id) {
      resourcesById[id] = entry.resource as Resource;
    }
  });

  data.forEach((entry) => {
    if (entry.search?.mode === MATCH_SEARCH_MODE) {
      const resource = entry.resource as Dictionary;
      const rowData = {
        key: resource.id,
        id: resource.id,
        name: resource.name,
        type: resource.physicalType?.coding[0]?.display,
        status: resource.status,
        parent: getResourceParentName(resource, resourcesById),
        location: resource,
      };
      tableData.push(rowData);
    }
  });
  return tableData;
}

export const eusmPhysicalLocationsFilterParams = {
  type: `${hl7PhysicalTypeCodeSystemUri}|${PhysicalTypeCodes.BUILDING}`,
};
