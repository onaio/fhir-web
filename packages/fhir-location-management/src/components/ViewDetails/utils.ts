import { IBundle } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/IBundle';
import { ILocation } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/ILocation';
import { locationResourceType } from '../../constants';
import {
  FHIRServiceClass,
  getObjLike,
  getResourcesFromBundle,
  isValidDate,
} from '@opensrp/react-utils';
import { get, isEqual, reverse } from 'lodash';
import { Coding } from '@smile-cdr/fhirts/dist/FHIR-R4/classes/coding';
import {
  eusmServicePointCodeSystemUri,
  getLocationAdmLevel,
  locationGeometryExtensionUri,
} from '@opensrp/fhir-helpers';
import { Extension } from '@smile-cdr/fhirts/dist/FHIR-R4/classes/extension';
import { CodeableConcept } from '@smile-cdr/fhirts/dist/FHIR-R4/classes/codeableConcept';

/**
 * fetches the location as well as all of its parent locations.
 *
 * @param fhirBaseUrl - fhir base url.
 * @param locationId - location of interest.
 */
export async function getLocationsAncestors(fhirBaseUrl: string, locationId: string) {
  let foundParent = false;
  const locationPath: ILocation[] = [];
  let currentLocationId = locationId;
  while (!foundParent) {
    const { parent, child } = await fetchLocAndParent(fhirBaseUrl, currentLocationId);
    if (child) {
      locationPath.push(child);
    }
    if (!parent) {
      foundParent = true;
    } else {
      locationPath.push(parent);
      const grandParentReferenceParts = (parent.partOf?.reference ?? '').split('/');
      const grandParentReference = grandParentReferenceParts[grandParentReferenceParts.length - 1];
      currentLocationId = grandParentReference;
      if (!grandParentReference) {
        foundParent = true;
      }
    }
  }
  const rootFirstLocations = reverse(locationPath);
  return rootFirstLocations;
}

/**
 * Fetches a location and its parent at a time.
 *
 * @param fhirBaseUrl - base url for the location.
 * @param locationId - id of the location of interest.
 */
async function fetchLocAndParent(fhirBaseUrl: string, locationId?: string) {
  const service = new FHIRServiceClass<IBundle>(fhirBaseUrl, locationResourceType);
  const response = await service
    .list({ _id: locationId, _include: 'Location:partof' })
    .then((res) => getResourcesFromBundle<ILocation>(res));
  // invariant we should only have at-most 2 locations here.
  let parentId;
  let parent: ILocation | undefined = undefined;
  let child: ILocation | undefined = undefined;
  for (const loc of response) {
    const thisLocParentId = loc.partOf?.reference;
    const thisLocId = `${locationResourceType}/${loc.id}`;
    if (parentId === undefined) {
      parentId = thisLocParentId;
    }
    if (parentId === thisLocId) {
      parent = loc;
    } else {
      child = loc;
    }
  }
  return { child, parent };
}

/**
 * parse a location resource into an obj for easier consumption
 *
 * @param node - the location resource
 */
export function parseLocationDetails(node: ILocation) {
  const { id, name, status, description, identifier, meta, extension, position, alias, type } =
    node;
  const updatedLastRaw = meta?.lastUpdated;

  let updatedLast = updatedLastRaw ? new Date(updatedLastRaw) : undefined;
  if (!isValidDate(updatedLast)) {
    updatedLast = undefined;
  }

  const geoJSonExtens = getObjLike(extension, 'url', locationGeometryExtensionUri)[0] as
    | Extension
    | undefined;
  const geoJsonExtensionData = geoJSonExtens?.valueAttachment?.data;
  let geometry;
  if (geoJsonExtensionData) {
    geometry = btoa(geoJsonExtensionData);
  }

  const typeCodingFlatMap = type?.flatMap((concept) => concept.coding) ?? [];
  const servicePointTypeCodings = typeCodingFlatMap.filter((x) => x !== undefined) as Coding[];
  const servicePointCode = getObjLike(
    servicePointTypeCodings,
    'system',
    eusmServicePointCodeSystemUri
  )[0] as Coding | undefined;

  const administrativeLevel = getLocationAdmLevel(type as CodeableConcept[]);

  return {
    identifier,
    id: id,
    name: name,
    partOf: node.partOf?.display ?? '-',
    description: description,
    status: status,
    physicalType: {
      code: get(node, 'physicalType.coding.0.code'),
      display: get(node, 'physicalType.coding.0.display'),
    },
    lastUpdated: updatedLast?.toLocaleString(),
    geometry,
    longitude: position?.longitude,
    latitude: position?.latitude,
    version: meta?.versionId,
    alias,
    servicePointType: servicePointCode?.display,
    administrativeLevel,
  };
}

export const hasCode = (codeList: Coding[], coding: Coding) => {
  for (const code of codeList) {
    if (isEqual(code, coding)) {
      return true;
    }
  }
  return false;
};
