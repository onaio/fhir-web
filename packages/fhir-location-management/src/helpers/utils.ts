import { ChildNodeList, LocationHierarchyResource, ParsedHierarchyNode, TreeNode } from './types';
import { cloneDeep } from 'lodash';
import cycle from 'cycle';
import TreeModel from 'tree-model';
import { IBundle } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/IBundle';
import { Resource } from '@smile-cdr/fhirts/dist/FHIR-R4/classes/resource';
import { UseQueryOptions, useQuery } from 'react-query';
import { FHIRServiceClass } from '@opensrp/react-utils';
import { locationHierarchyResourceType, locationResourceType } from '../constants';
import { ILocation } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/ILocation';
import { HTTPError } from '@opensrp/server-service';

/**
 * Parse the raw child hierarchy node map
 *
 * @param rawNodeMap - Object of raw hierarchy nodes
 * @returns Array of Parsed hierarchy nodes
 */
const parseFhirChildren = (rawNodeMap: ChildNodeList[]) => {
  return rawNodeMap.map((child) => {
    // standardize the parsedHierarchy node structure to be similar for both
    // root node and all other descendant nodes.
    const { treeNode } = child;
    const parsedNode: ParsedHierarchyNode = {
      ...treeNode,
      children: parseFhirChildren(treeNode.children ?? []) as ParsedHierarchyNode[],
    };
    return parsedNode;
  });
};

export const parseFHIRHierarchy = (fhirTree: LocationHierarchyResource) => {
  const rawClone: LocationHierarchyResource = cloneDeep(fhirTree);

  const { listOfNodes } = rawClone.LocationHierarchyTree.locationsHierarchy;
  const rawNode = listOfNodes.treeNode[0];

  const parsedNode = {
    ...rawNode,
    children: parseFhirChildren(rawNode.children ?? []),
  };
  return parsedNode;
};

export const generateFhirLocationTree = (rootNode: LocationHierarchyResource) => {
  const tree = new TreeModel();
  const hierarchy = parseFHIRHierarchy(rootNode);
  const root = tree.parse<ParsedHierarchyNode>(hierarchy);
  return root;
};

export const convertApiResToTree = (apiRes: IBundle) => {
  const rootNode = apiRes.entry?.[0].resource as Resource as LocationHierarchyResource | undefined;
  if (!rootNode) {
    return;
  }
  return generateFhirLocationTree(rootNode);
};

/**
 * serialize tree due to circular dependencies
 *
 * @param trees - trees to be serialized
 */
export const serializeTree = (trees?: TreeNode[] | TreeNode) => {
  if (!trees) {
    return JSON.stringify(undefined);
  }
  const sanitizeTrees = Array.isArray(trees) ? trees : [trees];
  return JSON.stringify(sanitizeTrees.map((tree) => JSON.stringify(cycle.decycle(tree))));
};

/**
 * get the location hierarchy of location with given identifier
 *
 * @param baseUrl - the server base url
 * @param rootId - the location identifier
 * @param queryOptions - extra query options.
 */
export const useGetLocationHierarchy = (
  baseUrl: string,
  rootId: string,
  queryOptions: UseQueryOptions<IBundle, HTTPError, TreeNode> = {}
) => {
  const hierarchyParams = {
    _id: rootId,
  };
  return useQuery<IBundle, HTTPError, TreeNode>(
    [locationHierarchyResourceType, hierarchyParams],
    async () => {
      return new FHIRServiceClass<IBundle>(baseUrl, locationHierarchyResourceType).list(
        hierarchyParams
      );
    },
    {
      select: (res: IBundle) => {
        return convertApiResToTree(res) as TreeNode;
      },
      refetchInterval: false,
      staleTime: Infinity,
      ...queryOptions,
    }
  );
};

/**
 * get single location
 *
 * @param baseUrl - the server base url
 * @param locId - the location identifier
 */
export const useGetLocation = (baseUrl: string, locId?: string) => {
  const serve = new FHIRServiceClass<ILocation>(baseUrl, locationResourceType);
  return useQuery([locationResourceType, locId], () => serve.read(locId as string), {
    select: (res) => res,
    enabled: !!locId,
  });
};
