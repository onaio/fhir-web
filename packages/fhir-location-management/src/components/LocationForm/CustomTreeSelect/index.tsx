import React, { useEffect } from 'react';
import { TreeNode } from '../../../helpers/types';
import { serializeTree } from '../../../helpers/utils';
import { TreeSelect } from 'antd';
import { TreeSelectProps } from 'antd/lib/tree-select/';
import { LabelValueType, DataNode } from 'rc-tree-select/lib/interface';
import { treeToOptions } from '../utils';

/** props for service types select component */
export interface CustomTreeSelectProps extends TreeSelectProps<LabelValueType> {
  tree: TreeNode;
  fullDataCallback?: (node: TreeNode) => void;
  disabledTreeNodesCallback?: (node: TreeNode) => boolean;
}

/**
 * form field where user can select the parent location from a tree structure
 *
 * @param props - the component props
 */
export const CustomTreeSelect = (props: CustomTreeSelectProps) => {
  const { value, fullDataCallback, disabledTreeNodesCallback, tree, ...restProps } = props;

  useEffect(() => {
    // if value is set, find parent node and pass it to fullDataCallback
    const node = tree.first((node) => node.model.nodeId === value);
    if (node) {
      fullDataCallback?.(node);
    } else {
      fullDataCallback?.(tree);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [serializeTree(tree), fullDataCallback, value]);

  const userDefinedRoots = tree.children ?? [];
  const selectOptions = treeToOptions(userDefinedRoots, disabledTreeNodesCallback);

  const treeSelectProps: TreeSelectProps<DataNode> = {
    ...restProps,
    treeData: selectOptions,
    value,
  };

  return <TreeSelect {...treeSelectProps}></TreeSelect>;
};
