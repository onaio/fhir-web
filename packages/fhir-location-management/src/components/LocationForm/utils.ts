import { Dictionary } from '@onaio/utils';
import { Rule } from 'rc-field-form/lib/interface';
import { TreeNode } from '../../helpers/types';
import { DataNode } from 'rc-tree-select/lib/interface';
import { v4 } from 'uuid';
import { get } from 'lodash';
import { IfhirR4 } from '@smile-cdr/fhirts';
import { ILocation } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/ILocation';
import { FHIRServiceClass } from '@opensrp/react-utils';
import { LocationUnitStatus } from '../../helpers/types';
import type { TFunction } from '@opensrp/i18n';

export type ExtraFields = Dictionary;

/** describes known fields that the form will have */
export interface LocationFormFields {
  id?: string;
  name: string;
  status: LocationUnitStatus;
  parentId?: string;
  description?: string;
  alias?: string;
  isJurisdiction: boolean;
}

interface BaseSetting {
  key: string;
  description: string;
  uuid: string;
  settingsId: string;
  settingIdentifier: string;
  settingMetadataId: string;
  v1Settings: false;
  resolveSettings: false;
  documentId: string;
  serverVersion: number;
}

/** describes a single settings object as received from location settings api */
export interface LocationSetting extends BaseSetting {
  label: string;
}

/** describes a single settings object as received from service types settings api */
export interface ServiceTypeSetting extends BaseSetting {
  value: string;
}

export const defaultFormField: LocationFormFields = {
  id: '',
  name: '',
  parentId: undefined,
  status: LocationUnitStatus.ACTIVE,
  isJurisdiction: true,
  description: undefined,
  alias: undefined,
};

/**
 * helps compute the default values of the location form field values
 *
 * @param location - the location unit
 * @param parentId - parent node id
 */
export const getLocationFormFields = (
  location?: IfhirR4.ILocation,
  parentId?: string
): LocationFormFields => {
  return {
    ...(location ?? defaultFormField),
    isJurisdiction: true, // TODO
    parentId: parentId ?? location?.partOf?.reference,
  } as LocationFormFields;
};

/**
 * util method to generate a location unit payload from form values
 *
 * @param formValues - values from the form
 * @param initialValues - initial values
 * @param parentNode - parent node of created location
 */
export const generateLocationUnit = (
  formValues: LocationFormFields,
  initialValues: LocationFormFields,
  parentNode?: TreeNode
) => {
  const { id, name, status, description, alias, isJurisdiction } = formValues;

  const uuid = get(initialValues, 'identifier.0.value');
  const thisLocationsIdentifier = uuid ? uuid : v4();

  let partOf: ILocation['partOf'];
  if (parentNode) {
    // this is a user defined location
    partOf = {
      reference: parentNode.model.nodeId,
      display: parentNode.model.node.name,
    };
  }

  const payload = {
    resourceType: 'Location',
    status,
    name,
    alias,
    description,
    partOf: partOf,
    identifier: [
      {
        use: 'official',
        value: thisLocationsIdentifier,
      },
    ],
    physicalType: {
      coding: [
        {
          system: 'http://terminology.hl7.org/CodeSystem/location-physical-type',
          code: isJurisdiction ? 'jdn' : 'bu',
          display: isJurisdiction ? 'Jurisdiction' : 'Building',
        },
      ],
    },
  } as ILocation;

  if (id) {
    payload.id = id;
  } else {
    payload.id = v4();
  }

  return payload;
};

/**
 * factory for validation rules for LocationForm component
 *
 * @param t - language translator
 */
export const validationRulesFactory = (t: TFunction) => ({
  id: [{ type: 'string' }] as Rule[],
  parentId: [
    { type: 'string', message: t(`Parent ID can only contain letters, numbers and spaces`) },
    {
      required: false,
    },
  ] as Rule[],
  name: [
    { type: 'string', message: t(`Name can only contain letters, numbers and spaces`) },
    { required: true, message: t(`Name is required`) },
  ] as Rule[],
  alias: [
    { type: 'string', message: t(`Name can only contain letters, numbers and spaces`) },
    { required: true, message: t(`Alias is required`) },
  ] as Rule[],
  status: [{ type: 'string' }, { required: true, message: t(`Status is required`) }] as Rule[],
  isJurisdiction: [
    {
      type: 'boolean',
    },
    {
      required: false,
    },
  ] as Rule[],
  description: [{ type: 'string' }, { required: false }] as Rule[],
});

/**
 * generates tree select options
 *
 * @param trees - an array of parsed trees
 * @param parentIdDisabledCallback - callback to help determine disabled status of nodes in treeSelect
 */
export const treeToOptions = (
  trees: TreeNode[],
  parentIdDisabledCallback?: (node: TreeNode) => boolean
): DataNode[] => {
  const recurseCreateOptions = (node: TreeNode) => {
    const optionValue: Dictionary = {
      value: node.model.nodeId,
      title: node.model.node.name,
      ...(parentIdDisabledCallback ? { disabled: parentIdDisabledCallback(node) } : {}),
    };
    if (node.hasChildren()) {
      optionValue.children = node.children.map(recurseCreateOptions);
    }
    return optionValue;
  };
  return trees.map(recurseCreateOptions);
};

/**
 * @param payload - the payload
 * @param baseUrl -  base url of api
 */
export async function postPutLocationUnit(payload: ILocation, baseUrl: string) {
  const serve = new FHIRServiceClass<ILocation>(baseUrl, 'Location');
  return serve.update(payload).catch((err: Error) => {
    throw err;
  });
}
