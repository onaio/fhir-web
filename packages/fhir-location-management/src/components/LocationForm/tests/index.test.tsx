import { store } from '@opensrp/store';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { Router } from 'react-router';
import { LocationForm } from '..';
import { getLocationFormFields } from '../utils';
import { createBrowserHistory } from 'history';
import { authenticateUser } from '@onaio/session-reducer';
import { Form } from 'antd';
import { createdLocation1, createdLocation2 } from './fixtures';
import nock, { RequestBodyMatcher } from 'nock';
import { QueryClientProvider, QueryClient } from 'react-query';
import { locationHierarchyResourceType } from '../CustomTreeSelect';
import { fhirHierarchy, onaOfficeSubLocation } from '../../../ducks/tests/fixtures';
import { convertApiResToTree } from '../../../helpers/utils';
import { cleanup, waitFor } from '@testing-library/react';
import flushPromises from 'flush-promises';
import * as notifications from '@opensrp/notifications';

jest.mock('@opensrp/notifications', () => ({
  __esModule: true,
  ...Object.assign({}, jest.requireActual('@opensrp/notifications')),
}));

const history = createBrowserHistory();
nock.disableNetConnect();

jest.mock('fhirclient', () => {
  return jest.requireActual('fhirclient/lib/entry/browser');
});

jest.mock('uuid', () => {
  const actual = jest.requireActual('uuid');
  return {
    ...actual,
    v4: () => '9b782015-8392-4847-b48c-50c11638656b',
  };
});

describe('LocationForm', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        cacheTime: 0,
      },
    },
  });

  const AppWrapper = (props: { children: React.ReactNode }) => {
    return (
      <Router history={history}>
        <QueryClientProvider client={queryClient}>{props.children}</QueryClientProvider>;
      </Router>
    );
  };

  const tree = convertApiResToTree(fhirHierarchy);
  const formProps = {
    fhirBaseURL: 'http://test.server.org',
    fhirRootLocationId: 'potus',
    tree,
  };

  beforeAll(() => {
    store.dispatch(
      authenticateUser(
        true,
        {
          email: 'bob@example.com',
          name: 'Bobbie',
          username: 'RobertBaratheon',
        },
        { api_token: 'hunter2', oAuth2Data: { access_token: 'sometoken', state: 'abcde' } }
      )
    );
  });

  afterEach(() => {
    nock.cleanAll();
    cleanup();
  });

  it('renders correctly', async () => {
    const div = document.createElement('div');
    document.body.appendChild(div);

    nock(formProps.fhirBaseURL)
      .get(`/${locationHierarchyResourceType}/_search`)
      .query({ _id: formProps.fhirRootLocationId })
      .reply(200, fhirHierarchy)
      .persist();

    const wrapper = mount(
      <AppWrapper>
        <LocationForm {...formProps} />
      </AppWrapper>,
      { attachTo: div }
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(toJson(wrapper.find('#parentId label'))).toMatchSnapshot('parentId label');
    expect(toJson(wrapper.find('#parentId input'))).toMatchSnapshot('parentId field');

    expect(toJson(wrapper.find('#name label'))).toMatchSnapshot('name label');
    expect(toJson(wrapper.find('#name input'))).toMatchSnapshot('name field');

    expect(toJson(wrapper.find('#status label').first())).toMatchSnapshot('status label');
    expect(toJson(wrapper.find('#status input'))).toMatchSnapshot('status field');

    expect(toJson(wrapper.find('#isJurisdiction label').first())).toMatchSnapshot(
      'isJurisdiction label'
    );
    expect(toJson(wrapper.find('#isJurisdiction input'))).toMatchSnapshot('isJurisdiction field');

    expect(toJson(wrapper.find('#description label'))).toMatchSnapshot('description label');
    expect(toJson(wrapper.find('#description textarea'))).toMatchSnapshot('description field');

    expect(toJson(wrapper.find('#location-form-submit-button button'))).toMatchSnapshot(
      'submit button'
    );
    expect(toJson(wrapper.find('#location-form-cancel-button button'))).toMatchSnapshot(
      'cancel button'
    );
    wrapper.unmount();
  });

  it('form validation works', async () => {
    const div = document.createElement('div');
    document.body.appendChild(div);

    nock(formProps.fhirBaseURL)
      .get(`/${locationHierarchyResourceType}/_search`)
      .query({ _id: formProps.fhirRootLocationId })
      .reply(200, fhirHierarchy)
      .persist();

    // when instance is set to core by default, types is required, serviceType is not required
    const wrapper = mount(
      <AppWrapper>
        <LocationForm {...formProps} />
      </AppWrapper>,
      { attachTo: div }
    );

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await flushPromises();
    });
    wrapper.update();

    await waitFor(() => {
      const atLeastOneError = document.querySelector('.ant-form-item-explain-error');
      expect(atLeastOneError).toBeInTheDocument();
    });

    // not required
    expect(wrapper.find('#parentId .ant-form-item').text()).toMatchInlineSnapshot(
      `"Part OfSelect the parent location"`
    );

    // name is required and has not default
    expect(wrapper.find('#name .ant-form-item').text()).toMatchInlineSnapshot(
      `"NameName is required"`
    );

    // status has default value
    expect(wrapper.find('#status .ant-form-item').text()).toMatchInlineSnapshot(
      `"StatusActiveInactive"`
    );

    // has default value
    expect(wrapper.find('#isJurisdiction .ant-form-item').text()).toMatchInlineSnapshot(
      `"Physical typeJurisdictionBuilding"`
    );

    expect(wrapper.find('#alias .ant-form-item').text()).toMatchInlineSnapshot(`"Alias"`);

    // not required
    expect(wrapper.find('#description .ant-form-item').text()).toMatchInlineSnapshot(
      `"Description"`
    );

    wrapper.unmount();
  });

  it('submits new location unit', async () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    nock(formProps.fhirBaseURL)
      .put(`/Location/${createdLocation1.id}`, createdLocation1 as unknown as RequestBodyMatcher)
      .reply(201, {})
      .persist();

    const someMockURL = '/someURL';
    const successURLGeneratorMock = jest.fn(() => someMockURL);
    const notificationSuccessMock = jest.spyOn(notifications, 'sendSuccessNotification');

    const wrapper = mount(
      <AppWrapper>
        <LocationForm successURLGenerator={successURLGeneratorMock} {...formProps} />
      </AppWrapper>,
      { attachTo: container }
    );

    await act(async () => {
      await flushPromises();
    });
    wrapper.update();

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const formInstance = (wrapper.find(Form).props() as any).form;

    // change parent Id
    formInstance.setFieldsValue({
      parentId: 'Location/303',
    });

    wrapper.update();

    // simulate active check to be suspended
    wrapper
      .find('#status .ant-form-item input')
      .last()
      .simulate('change', {
        target: { checked: true },
      });

    // set isJurisdiction to structure
    wrapper
      .find('#isJurisdiction .ant-form-item input')
      .first()
      .simulate('change', {
        target: { checked: true },
      });

    // simulate name change
    wrapper
      .find('#name .ant-form-item input')
      .simulate('change', { target: { name: 'name', value: 'area51' } });

    wrapper
      .find('#alias .ant-form-item input')
      .simulate('change', { target: { name: 'alias', value: 'creepTown' } });

    wrapper.find('#description .ant-form-item textarea').simulate('change', {
      target: {
        value:
          'The secret Nevada base, known as Area 51, is often the subject of alien conspiracy theories.',
      },
    });

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await flushPromises();
    });

    await waitFor(() => {
      expect(notificationSuccessMock).toHaveBeenCalledWith('Location was successfully created');
    });
    wrapper.update();
    wrapper.unmount();
  });

  it('cancel handler is called on cancel', async () => {
    const container = document.createElement('div');
    document.body.appendChild(container);

    nock(formProps.fhirBaseURL)
      .get(`/${locationHierarchyResourceType}/_search`)
      .query({ _id: formProps.fhirRootLocationId })
      .reply(200, fhirHierarchy)

      .persist();

    const cancelMock = jest.fn();

    const wrapper = mount(
      <AppWrapper>
        <LocationForm onCancel={cancelMock} {...formProps} />
      </AppWrapper>,

      { attachTo: container }
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    wrapper.find('button#location-form-cancel-button').simulate('click');
    wrapper.update();

    expect(cancelMock).toHaveBeenCalled();
    wrapper.unmount();
  });

  it('is able to edit a location unit', async () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const notificationSuccessMock = jest.spyOn(notifications, 'sendSuccessNotification');

    nock(formProps.fhirBaseURL)
      .get(`/${locationHierarchyResourceType}/_search`)
      .query({ _id: formProps.fhirRootLocationId })
      .reply(200, fhirHierarchy)
      .put(`/Location/${createdLocation2.id}`, createdLocation2 as unknown as RequestBodyMatcher)
      .reply(201, {})
      .persist();

    const initialValues = getLocationFormFields(onaOfficeSubLocation);

    const wrapper = mount(
      <AppWrapper>
        <LocationForm initialValues={initialValues} {...formProps} />
      </AppWrapper>,
      { attachTo: container }
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    // simulate active check to be active
    wrapper
      .find('#status .ant-form-item input')
      .first()
      .simulate('change', {
        target: { checked: true },
      });

    // set isJurisdiction to structure
    wrapper
      .find('#isJurisdiction .ant-form-item input')
      .last()
      .simulate('change', {
        target: { checked: true },
      });

    // simulate name change
    wrapper
      .find('#name .ant-form-item input')
      .simulate('change', { target: { name: 'name', value: 'Yosemite' } });

    wrapper
      .find('#alias .ant-form-item input')
      .simulate('change', { target: { name: 'alias', value: 'world wonder' } });

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    await waitFor(() => {
      expect(notificationSuccessMock).toHaveBeenCalledWith('Location was successfully updated');
    });

    wrapper.unmount();
  });
});
