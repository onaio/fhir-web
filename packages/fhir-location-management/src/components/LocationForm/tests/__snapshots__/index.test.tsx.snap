// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationForm renders correctly: cancel button 1`] = `
<button
  className="ant-btn css-dev-only-do-not-override-k7429z ant-btn-default"
  disabled={false}
  id="location-form-cancel-button"
  onClick={[Function]}
  type="button"
>
  <LoadingIcon
    existIcon={false}
    loading={false}
    prefixCls="ant-btn"
  >
    <CSSMotion
      motionName="ant-btn-loading-icon-motion"
      onAppearActive={[Function]}
      onAppearStart={[Function]}
      onEnterActive={[Function]}
      onEnterStart={[Function]}
      onLeaveActive={[Function]}
      onLeaveStart={[Function]}
      removeOnLeave={true}
      visible={false}
    >
      <DomWrapper />
    </CSSMotion>
  </LoadingIcon>
  <span
    key=".0"
  >
    Cancel
  </span>
</button>
`;

exports[`LocationForm renders correctly: description field 1`] = `
<textarea
  className="ant-input css-dev-only-do-not-override-k7429z"
  disabled={false}
  hidden={null}
  id="location-form_description"
  onBlur={[Function]}
  onChange={[Function]}
  onCompositionEnd={[Function]}
  onCompositionStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  placeholder="Description"
  rows={4}
  style={
    Object {
      "resize": undefined,
    }
  }
  value=""
/>
`;

exports[`LocationForm renders correctly: description label 1`] = `
<label
  className=""
  htmlFor="location-form_description"
  title="Description"
>
  Description
</label>
`;

exports[`LocationForm renders correctly: isJurisdiction field 1`] = `
Array [
  <input
    checked={true}
    className="ant-radio-input"
    disabled={false}
    onChange={[Function]}
    type="radio"
    value={true}
  />,
  <input
    checked={false}
    className="ant-radio-input"
    disabled={false}
    onChange={[Function]}
    type="radio"
    value={false}
  />,
]
`;

exports[`LocationForm renders correctly: isJurisdiction label 1`] = `
<label
  className=""
  htmlFor="location-form_isJurisdiction"
  title="Physical type"
>
  Physical type
</label>
`;

exports[`LocationForm renders correctly: name field 1`] = `
<input
  aria-required="true"
  className="ant-input css-dev-only-do-not-override-k7429z"
  disabled={false}
  hidden={null}
  id="location-form_name"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  placeholder="Enter a location name"
  type="text"
  value=""
/>
`;

exports[`LocationForm renders correctly: name label 1`] = `
<label
  className="ant-form-item-required"
  htmlFor="location-form_name"
  title="Name"
>
  Name
</label>
`;

exports[`LocationForm renders correctly: parentId field 1`] = `
<input
  aria-autocomplete="list"
  aria-controls="location-form_parentId_list"
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="location-form_parentId_list"
  autoComplete="off"
  className="ant-select-selection-search-input"
  disabled={false}
  id="location-form_parentId"
  onChange={[Function]}
  onCompositionEnd={[Function]}
  onCompositionStart={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onPaste={[Function]}
  readOnly={true}
  role="combobox"
  style={
    Object {
      "opacity": 0,
    }
  }
  type="search"
  unselectable="on"
  value=""
/>
`;

exports[`LocationForm renders correctly: parentId label 1`] = `
<label
  className=""
  htmlFor="location-form_parentId"
  title="Part Of"
>
  Part Of
</label>
`;

exports[`LocationForm renders correctly: status field 1`] = `
Array [
  <input
    checked={true}
    className="ant-radio-input"
    disabled={false}
    name="active"
    onChange={[Function]}
    type="radio"
    value="active"
  />,
  <input
    checked={false}
    className="ant-radio-input"
    disabled={false}
    name="active"
    onChange={[Function]}
    type="radio"
    value="inactive"
  />,
]
`;

exports[`LocationForm renders correctly: status label 1`] = `
<label
  className="ant-form-item-required"
  htmlFor="location-form_status"
  title="Status"
>
  Status
</label>
`;

exports[`LocationForm renders correctly: submit button 1`] = `
<button
  className="ant-btn css-dev-only-do-not-override-k7429z ant-btn-primary"
  disabled={false}
  id="location-form-submit-button"
  onClick={[Function]}
  type="submit"
>
  <LoadingIcon
    existIcon={false}
    loading={false}
    prefixCls="ant-btn"
  >
    <CSSMotion
      motionName="ant-btn-loading-icon-motion"
      onAppearActive={[Function]}
      onAppearStart={[Function]}
      onEnterActive={[Function]}
      onEnterStart={[Function]}
      onLeaveActive={[Function]}
      onLeaveStart={[Function]}
      removeOnLeave={true}
      visible={false}
    >
      <DomWrapper />
    </CSSMotion>
  </LoadingIcon>
  <span
    key=".0"
  >
    Save
  </span>
</button>
`;
